### PR TITLE
Add prelu op and module for quantized CPU backend

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
Add prelu op and module for quantized CPU backend.
The PR includes:
- Quantized version of prelu op
- Native prelu kernel for quantized CPU
- Prelu modules in `nn` and `nn.quantized`
- FX support for prelu
- Unit tests

### CODE REVIEW

**Comment by jerryzh168** [comment]
please rebase on master, there is some code updates

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168. There is one check failure which says `torch.nn.quantized.functional.prelu` is undocumented. What should I do to document it?

**Comment by jerryzh168** [comment]
hmmm, we don't need nn.functional.prelu I think, we should just remove it, we just need torch.ops.quantized.prelu, for documentations, I think @vkuzo is taking a look recently, not sure how we want to do it.

**Comment by Xia-Weiwen** [comment]
Thanks. I have removed `nn.functional.prelu`.
Please review again. Thanks.

**Comment by jerryzh168** [comment]
if the weight is float, what is quantized here?

**Comment by Xia-Weiwen** [comment]
Input and output are quantized. Please see the comments by @mingfeima below. Thanks.

https://github.com/pytorch/pytorch/pull/73491#issuecomment-1056036791
> shall we keep `weight` in float when converting `prelu` to `qprelu`? kind of like [`qbatch_norm`](https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/batchnorm.py#L14) here.
> 
> Since we don't have native assembly to handle it (need to convert to float and then compute)...
> 
> To quantize weight doesn't make much sense since: a) lose both performance and accuracy b) weight is usually of small memory footprint so quantizing it won't save much.

File: torch/ao/quantization/fx/quantization_patterns.py [hunk 1]
```diff
@@ -243,6 +243,8 @@ def convert(self,
     torch.nn.functional.softmax: int8_dtypes,
     torch.nn.functional.dropout: int8_dtypes,
     torch.sum: fp16_dtypes,
+    torch.nn.PReLU: int8_dtypes,
+    torch.nn.functional.prelu: int8_dtypes,
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/quantization_patterns.py:247]: this is deprecated I think, we have some refactoring going on right now and new ops need to be added in torch/ao/quantization/fx/backend_config/native.py

File: torch/ao/quantization/fx/quantization_patterns.py [hunk 2]
```diff
@@ -1312,6 +1314,7 @@ def convert(self,
 @register_quant_pattern(torch.nn.SiLU)
 @register_quant_pattern(torch.nn.Mish)
 @register_quant_pattern(torch.nn.Dropout)
+@register_quant_pattern(torch.nn.PReLU)
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/quantization_patterns.py:1317]: same for these

**Comment by jerryzh168** [comment]
> Input and output are quantized. Please see the comments by @mingfeima below. Thanks.
> 
> [#73491 (comment)](https://github.com/pytorch/pytorch/pull/73491#issuecomment-1056036791)
> 
> > shall we keep `weight` in float when converting `prelu` to `qprelu`? kind of like [`qbatch_norm`](https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/batchnorm.py#L14) here.
> > Since we don't have native assembly to handle it (need to convert to float and then compute)...
> > To quantize weight doesn't make much sense since: a) lose both performance and accuracy b) weight is usually of small memory footprint so quantizing it won't save much.

thanks, sounds good. looks like we use dequantize - operation - quantize in kernels, do we have any plans to improve the perf with more efficient implementations? maybe at that time we also need to quantize the weight so that computation can happen in int8?

**Comment by mingfeima** [comment]
> > Input and output are quantized. Please see the comments by @mingfeima below. Thanks.
> > [#73491 (comment)](https://github.com/pytorch/pytorch/pull/73491#issuecomment-1056036791)
> > > shall we keep `weight` in float when converting `prelu` to `qprelu`? kind of like [`qbatch_norm`](https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/batchnorm.py#L14) here.
> > > Since we don't have native assembly to handle it (need to convert to float and then compute)...
> > > To quantize weight doesn't make much sense since: a) lose both performance and accuracy b) weight is usually of small memory footprint so quantizing it won't save much.
> 
> thanks, sounds good. looks like we use dequantize - operation - quantize in kernels, do we have any plans to improve the perf with more efficient implementations? maybe at that time we also need to quantize the weight so that computation can happen in int8?

@jerryzh168 , good idea:) we should be able to implement something like [qmul_kernel](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L1290) with `ReluFused = True`, this should be more efficient. @Xia-Weiwen is working on it.

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 3]
```diff
@@ -616,6 +616,116 @@ static void leaky_qrelu_out_kernel(Tensor& out, const Tensor& qx,
...
+
+    qVec in_zp_vec = qVec(static_cast<scalar_t>(in_zero_point));
+    qVec wei_zp_vec = qVec(static_cast<scalar_t>(wei_zero_point));
+
+    qVec scalar_wei_vec = qVec(static_cast<scalar_t>(W_ptr[0]));
+    auto quantized_one = at::native::quantize_val<scalar_t>(wei_scale, wei_zero_point, 1.0f);
+    qVec one_qvec = qVec(quantized_one);
+    auto one_qvec_sub_zp = one_qvec.widening_subtract(wei_zp_vec)[0];
+    int32_t one_sub_zp = quantized_one.val_ - static_cast<int32_t>(wei_zero_point);
+
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Hi @mingfeima I have reimplemented it with `cpu_kernel_vec`. Could you review again? Thanks.

File: aten/src/ATen/native/quantized/cpu/qrelu.cpp [hunk 4]
```diff
@@ -134,6 +135,64 @@ Tensor& leaky_relu_quantized_cpu_(Tensor& self, const Scalar& negval) {
...
+    qx = qx.unsqueeze(-1);
+  } else if (ndim == 5) {
+    format = MemoryFormat::ChannelsLast3d;
+  }
+  qx = qx.contiguous(format);
+
+  auto qy = at::_empty_affine_quantized(qx.sizes(),
+      at::device(kCPU)
+        .dtype(self.scalar_type())
+        .memory_format(format),
```
Comment (mingfeima) [inline aten/src/ATen/native/quantized/cpu/qrelu.cpp:170]: create with `qx.suggest_memory_format()`, we can't fix it here.

File: aten/src/ATen/native/quantized/cpu/qrelu.cpp [hunk 5]
```diff
@@ -134,6 +135,64 @@ Tensor& leaky_relu_quantized_cpu_(Tensor& self, const Scalar& negval) {
...
+  auto ndim = self.dim();
+  // for ndim < 1 or > 5, go to reference path
+  if (ndim > 5 || ndim < 1) {
+    auto x = self.dequantize();
+    auto y = at::prelu(x, weight);
+    return at::quantize_per_tensor(y, output_scale, output_zero_point, c10::kQUInt8);
+  }
+
+  // Create fake dimensions if input ndim in [1, 3]
+  Tensor qx = self;
```
Comment (mingfeima) [inline aten/src/ATen/native/quantized/cpu/qrelu.cpp:154]: we don't need to create a Channels last memory view upon qx here.

TensorIterator will re-calculate the shape of input tensor according to order of strides (re-arrange the dimension order from strides lowest to highest). AND memory format is only a tag informing the strides, so it makes no difference TensorIterator how to deal with it. 

only need to create a view upon `weight` so as to make sure that it's size is able to be broadcasted to input's size.

File: aten/src/ATen/native/quantized/cpu/qrelu.cpp [hunk 6]
```diff
@@ -134,6 +135,64 @@ Tensor& leaky_relu_quantized_cpu_(Tensor& self, const Scalar& negval) {
...
+  if (ndim == 1) {
+    qx = qx.unsqueeze(0).unsqueeze(-1).unsqueeze(-1);
+  } else if (ndim == 2) {
+    qx = qx.unsqueeze(-1).unsqueeze(-1);
+  } else if (ndim == 3) {
+    qx = qx.unsqueeze(-1);
+  } else if (ndim == 5) {
+    format = MemoryFormat::ChannelsLast3d;
+  }
+  qx = qx.contiguous(format);
```
Comment (mingfeima) [inline aten/src/ATen/native/quantized/cpu/qrelu.cpp:165]: this is not needed, TensorIterator will pick on the most inner dimension (if contiguous) to vectorize.

suppose we have a input which is a sub tensor of a bigger one, this is a performance downgrade.

File: aten/src/ATen/native/quantized/cpu/qrelu.cpp [hunk 7]
```diff
@@ -134,6 +135,64 @@ Tensor& leaky_relu_quantized_cpu_(Tensor& self, const Scalar& negval) {
...
+      at::device(kCPU)
+        .dtype(self.scalar_type())
+        .memory_format(format),
+      output_scale,
+      output_zero_point,
+      c10::nullopt);
+
+  qprelu_stub(self.device().type(), qy, qx, weight);
+
+  // Remove fake dimensions, and go back to contiguous format
```
Comment (mingfeima) [inline aten/src/ATen/native/quantized/cpu/qrelu.cpp:177]: not needed.

* qy is created w.r.t qx's shape and memory format
* make sure w is able to be broadcasted to qx
* feed qx and qy to TensorIterator

File: aten/src/ATen/native/quantized/cpu/qrelu.cpp [hunk 8]
```diff
@@ -134,6 +135,64 @@ Tensor& leaky_relu_quantized_cpu_(Tensor& self, const Scalar& negval) {
...
+  if (ndim < 4) {
+    auto result = qy.contiguous(MemoryFormat::Contiguous);
+    if (ndim == 1) {
+      result = result.squeeze(0).squeeze(-1).squeeze(-1);
```
Comment (mingfeima) [inline aten/src/ATen/native/quantized/cpu/qrelu.cpp:181]: Additionally, be careful when using all the view operators, e.g. squeeze, unsqueeze, narrow etc. This doesn't allocate storage but it creates Tensor meta info.

**Comment by Xia-Weiwen** [comment]
Hi @mingfeima @jerryzh168, thanks for your review. I have made changes per your comments. Could you please review again? Thanks.

File: test/quantization/fx/test_quantize_fx.py [hunk 9]
```diff
@@ -5576,6 +5576,28 @@ def test_elu(self):
     def test_leaky_relu(self):
         self._test_activation_impl(nn.LeakyReLU, F.leaky_relu, nnq.LeakyReLU, torch.ops.quantized.leaky_relu)
 
+    def test_prelu(self):
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:5579]: can we reuse `_test_activation_impl` here?

File: torch/nn/quantized/modules/activation.py [hunk 10]
```diff
@@ -183,3 +183,53 @@ def from_float(mod):
...
+    r"""This is the quantized equivalent of :class:`~torch.nn.PReLU`.
+
+    Args:
+        scale: quantization scale of the output tensor
+        zero_point: quantization zero point of the output tensor
+        num_parameters: number of parameters: 1, or the number of channels at input. Default: 1
+    """
+    def __init__(self, output_scale: float, output_zero_point: int,
+                 num_parameters: int = 1) -> None:
+        super(PReLU, self).__init__()
```
Comment (jerryzh168) [inline torch/nn/quantized/modules/activation.py:197]: nit: `super().__init__()`

**Review by jerryzh168 — APPROVED** [review]
Looks good!

File: test/quantization/fx/test_quantize_fx.py (hunk 9 above)
Comment (Xia-Weiwen) [reply test/quantization/fx/test_quantize_fx.py:5579]: Thanks for the suggestion. PReLU needs additional arguments since it has weight. For example, `num_parameters` for module constructor and `weight` for `forward` of functional op. How do you suggest adding them? By adding `**kwargs` or `*args` to `_test_activation_impl`? Then we need to do some `if ... else ...` to accommodate PReLU. Do we have to consider other activations with additional parameters?

File: torch/nn/quantized/modules/activation.py (hunk 10 above)
Comment (Xia-Weiwen) [reply torch/nn/quantized/modules/activation.py]: Hi @jerryzh168 I have rebased the PR and fixed this issue. But for the `prelu_quantized_cpu` you mentioned above, I am not sure how to reuse that. It needs additional arguments. Could you give some suggestions? Thanks!

Comment (Xia-Weiwen) [reply torch/nn/quantized/modules/activation.py]: Hi @jerryzh168 This PR is merged. If you think reuse of `_test_activation_impl` is still a problem, please let me know. Thanks.

**Comment by Xia-Weiwen** [comment]
Hi there. Shall we try merging this again?

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 11]
```diff
@@ -652,6 +652,81 @@ static void leaky_qrelu_out_kernel(Tensor& out, const Tensor& qx,
...
+  int64_t input_ndim = qx.dim();
+  TORCH_CHECK(input_ndim > 0, "qprelu: zero-dim input tensor is not allowed.");
+
+  // Helper to convert 1d tensors or scalar tensor to an nd tensor that broadcasts with input
+  // All elements go into the channel dimension
+  DimVector sizes(input_ndim, 1), strides(input_ndim, 0);
+  auto as_nd = [&](const Tensor& t) {
+    TORCH_INTERNAL_ASSERT(t.defined() && (t.dim() == 1 || t.dim() == 0));
+    sizes[1] = t.dim() == 1 ? t.sizes()[0] : 1;
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:678]: cc @Xia-Weiwen wondering if you have tested this locally or not, looks like this can be index out of range error when the input is one dimensional

Comment (Xia-Weiwen) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:678]: Hi @jerryzh168 @z-a-f Thanks for pointing out this flaw and sorry for the negligence.
The following UT has covered the case where `input_ndim=1`: https://github.com/pytorch/pytorch/blob/b472852778eef777ac182e2efcc008429f5fc4fe/test/quantization/core/test_quantized_op.py#L474
Unfortunately, it fails to discover this flaw. I guess it is due to the implementation of `DimVector`.
I have created a PR to fix this: https://github.com/pytorch/pytorch/pull/82393. Please review. Thanks!
